android/expoview/src/main/java/host/exp/exponent/notifications/managers/SchedulerManager.java

```diff
@@ -0,0 +1,152 @@
+
+  private Context mApplicationContext;
+
+  SchedulerManager(Context applicationContext) {
+    mApplicationContext = applicationContext;
+  }
+
+  @Override
+  public void triggerAll(String action) {
+    fetchSchedulersMap();
+
+    cancelAlreadyScheduled(null);
+
+    ArrayList<String> unsuccessful = new ArrayList<String>();
+
+    for (Map.Entry<String, SchedulerInterface> scheduler : mSchedulersMap.entrySet()) {
+      try {
+        scheduler.getValue().schedule(action);
+      } catch (UnableToScheduleException e) {
+        unsuccessful.add(scheduler.getKey());
+      }
+    }
+
+    for (String key : unsuccessful) {
+      this.removeScheduler(key);
+    }
+  }
+
+  @Override
+  public void removeAll(String experienceId) {
+    fetchSchedulersMap();
+    cancelAlreadyScheduled(experienceId);
+
+    ArrayList<String> toRemove = new ArrayList<String>();
+
+    for (Map.Entry<String, SchedulerInterface> scheduler : mSchedulersMap.entrySet()) {
+      if (experienceId == null || scheduler.getValue().getOwnerExperienceId().equals(experienceId)) {
+        scheduler.getValue().remove();
+        toRemove.add(scheduler.getKey());
+      }
+    }
+
+    for (String key : toRemove) {
+      mSchedulersMap.remove(key);
+    }
+  }
+
+  @Override
+  public void cancelAlreadyScheduled(String experienceId) {
+    fetchSchedulersMap();
+    for (SchedulerInterface scheduler : mSchedulersMap.values()) {
+      if (experienceId == null || scheduler.getOwnerExperienceId().equals(experienceId)) {
+        scheduler.cancel();
+      }
+    }
+  }
+
+  @Override
+  public void rescheduleOrDelete(String id) {
+    fetchSchedulersMap();
+    SchedulerInterface scheduler = mSchedulersMap.get(id);
+    if (scheduler == null) {
+      return;
+    }
+
+    if (!scheduler.canBeRescheduled()) {
+      this.removeScheduler(id);
+    } else {
+      try {
+        scheduler.schedule(null);
+      } catch (UnableToScheduleException e) {
+        this.removeScheduler(id);
+      }
+    }
+  }
+
+  @Override
+  public void removeScheduler(String id) {
+    fetchSchedulersMap();
+    SchedulerInterface scheduler = mSchedulersMap.get(id);
+    if (scheduler == null) {
+      return;
+    }
+    mSchedulersMap.remove(id);
+    scheduler.cancel();
+    scheduler.remove();
+  }
+
+  @Override
+  public void addScheduler(SchedulerInterface scheduler, Function<String, Boolean> handler) {
+    fetchSchedulersMap();
+
+    scheduler.setApplicationContext(mApplicationContext);
+    String id = scheduler.saveAndGetId();
+    mSchedulersMap.put(id, scheduler);
+    try {
+      scheduler.schedule(null);
+    } catch (UnableToScheduleException e) {
+      this.removeScheduler(id);
+      id = null;
+    }
+    handler.apply(id);
+  }
+
+  private List<Class> getSchedulerClasses() {
+    return Arrays.asList(CalendarSchedulerModel.class, IntervalSchedulerModel.class);
+  }
+
+  private void fetchSchedulersMap() {
+    if (!mFetchedFromDB) {
```

Comment:
This also looks like a good piece of code to be extracted. How about creatins something like SchedulersRepository? Which would return SchedulersMap and maybe create them (which might be used in Module? )